scale-decode-derive/src/lib.rs

```diff
@@ -0,0 +1,561 @@
+                type Value<'scale, 'info> = #path_to_type #ty_generics;
+
+                fn visit_composite<'scale, 'info>(
+                    self,
+                    value: &mut Composite<'scale, 'info>,
+                    type_id: TypeId,
+                ) -> Result<Self::Value<'scale, 'info>, Self::Error> {
+                    #visit_composite_body
+                }
+                fn visit_tuple<'scale, 'info>(
+                    self,
+                    value: &mut Tuple<'scale, 'info>,
+                    type_id: TypeId,
+                ) -> Result<Self::Value<'scale, 'info>, Self::Error> {
+                    #visit_tuple_body
+                }
+            }
+
+            impl #impl_generics DecodeAsFields for #path_to_type #ty_generics #where_clause  {
+                fn decode_as_fields(input: &mut &[u8], fields: &[PortableField], types: &PortableRegistry) -> Result<Self, Error> {
+                    let path = Default::default();
+                    let mut composite = Composite::new(input, &path, fields, types);
+                    let val = <#path_to_type #ty_generics>::into_visitor().visit_composite(&mut composite, TypeId(0))?;
+                    // Consume any remaining bytes and update input:
+                    composite.skip_decoding()?;
+                    *input = composite.bytes_from_undecoded();
+                    Ok(val)
+                }
+            }
+        };
+    )
+}
+
+// Given some named fields, generate impls like `field_name: get_field_value()` for each field. Do this for the compoiste and tuple impls.
+fn named_field_keyvals<'f>(
+    path_to_scale_decode: &'f syn::Path,
+    fields: &'f syn::FieldsNamed,
+) -> (usize, impl Iterator<Item = TokenStream2> + 'f, impl Iterator<Item = TokenStream2> + 'f) {
+    let field_keyval_impls = fields.named.iter().map(move |f| {
+        let field_attrs = FieldAttrs::from_attributes(&f.attrs).unwrap_or_default();
+        let field_ident = f.ident.as_ref().expect("named field has ident");
+        let field_name = field_ident.to_string();
+        let skip_field = field_attrs.skip;
+
+        // If a field is skipped, we expect it to have a Default impl to use to populate it instead.
+        if skip_field {
+            return (
+                false,
+                quote!(#field_ident: ::std::default::Default::default()),
+                quote!(#field_ident: ::std::default::Default::default())
+            )
+        }
+
+        (
+            true,
+            quote!(#field_ident: {
+                let val = *vals
+                    .get(&Some(#field_name))
+                    .ok_or_else(|| #path_to_scale_decode::Error::new(#path_to_scale_decode::error::ErrorKind::CannotFindField { name: #field_name.to_owned() }))?;
+                val.decode_as_type().map_err(|e| e.at_field(#field_name))?
+            }),
+            quote!(#field_ident: {
+                let val = vals.next().unwrap()?;
```

Comment:
nit: Would it be possible for this unwrap to panic? Maybe an `except("Cannot get next tuple field")`?

Reply:
It shouldn't be, because this impl is only valid when pulling fields from a scale encoded tuple-type, and so before this line we'll have already checked that the num fields lines up. But I agree; I changed them to "expects" so that if I messed up, at least it'll be more obvious what the issue is :)